### ISSUE TITLE
[WIP] New field type to allow streams of mixed-type content

### ISSUE THREAD

**Issue by gasman** [torchbox/wagtail#823]
Spec here (NB 'TumblrField' now has the proposed name StreamField):
https://docs.google.com/document/d/1eAu5zJ6j7JIxUEQOT9amWjjmG-RUq_LQ3vyc4sZCs7Y/edit?usp=sharing

## Additional technical background

We will partially re-implement the edit handler mechanism - the ad-hoc design of edit handlers has led to it being overly complex, and developing it further to support stream fields would incur technical debt. In this phase of development, this partial implementation does not need to deal with modelforms / formsets, since all data will be serialised to a single JSON field.

There will be three 'patterns' that allow blocks to be composed into more complex arrangements:

Struct: A fixed set of sub-blocks of possibly differing types, such as a Person block consisting of 'first name', 'surname' and 'photo' fields.

List: An arbitrary number of blocks all of the same type, which can be added to (via an 'add another' button), deleted and reordered.

Stream: An arbitrary number of blocks, each of which is one of several available types. Between each block, a menu bar is displayed for inserting an additional block at that point. Blocks can also be deleted and reordered.

In the above descriptions, a 'block' may be a simple field (such as a text field or image picker) or one of the above patterns: e.g. it's possible to have a list of structs, or a struct where one of the members is a list of fields.

## Breakdown of tasks

* [x] Mechanism for dynamically adding blocks to the edit page, supporting arbitrarily complex javascript behaviour within each block (including blocks that recursively support adding sub-blocks dynamically, such as Sir Trevor's list type)
* [x] A generalised 'sequence' type to handle the adding/reordering behaviour which is common to List and Stream, but without dictating a specific UI pattern. (Needs to be flexible enough to cope with e.g. keeping Stream's menu bars in the right position relative to the blocks being moved)
* [x] Menu logic for Stream
* [x] Ability to define specific Struct/List/Stream types by subclassing those types, much like Django's Form class
* [x] Implementing primitive block types such as text fields / choosers. (Hopefully we can implement blocks via the same API as Django form fields, which would make this unnecessary.)
* [ ] A rich text component restricted to simple text formatting (probably a stripped-down hallo.js in this phase)
* [x] A field validation mechanism
* [x] A serialisation mechanism
* [x] Front-end rendering of fields
* [x] ~~A `{% streamfield %}` template tag as per spec~~ (just use if/else instead)

**Comment by gasman** [comment]
Have now created `feature/streamfield` branches in the wagtail / wagtaildemo repos, for the task of transplanting the logic from https://github.com/gasman/wagtailstreamfield into Wagtail itself. To use it, check out the `feature/streamfield` branch of both repos, and run the following SQL:

    update demo_blogpage set body = '[
        {"type": "heading", "value": "The largest event for the things that the event is about!"},
        {"type": "image", "value": 42},
        {"type": "heading", "value": "Earlyish Bird tickets available now"},
        {"type": "speaker", "value": {"name": "Tim Berners-Lee", "job_title": "Web developer", "nicknames": ["Timmy", "Bernie"]}},
        {"type": "image", "value": 99}
    ]' where page_ptr_id=16;
    delete from wagtailcore_pagerevision where page_id = 16;

This converts the 'body' field of BlogPost to a StreamField, and populates page ID 16 (title "Blog Post") with demo content.

Currently implemented:

* Round-trip of data from the database to the edit form and back again
* Structural blocks (stream, struct, list) including nesting to arbitrary depth
* Addition / removal of list / stream items
* Attaching Javascript handlers of arbitrary complexity

Not yet implemented:

* Styling
* Reordering of list / stream items
* Front-end rendering
* Validation / error reporting
* Proper block types for image/document/snippet chooser, rich text, and anything else that FieldBlock doesn't cover
* Making StreamFields work inside InlinePanels (not tested, but I suspect the use of `__PREFIX__` strings will clash - probably fixable by changing to some other placeholder string)

**Comment by edrex** [comment]
Trying to test this out and stuck:

 - checked out `wagtaildemo@feature/streamfield-frontend`
 - `vagrant up`
 - `pip install --src ./deps -e git+https://github.com/torchbox/wagtail.git@feature/streamfield-frontend#egg=wagtail`

Error running `dj migrate`:

```
ValueError: Cannot alter field demo.BlogPage.body into demo.BlogPage.body
- they do not properly define db_type (are you using PostGIS 1.5 or badly-written
custom fields?)
```

Running the SQL you posted results in:

```
ERROR:  relation "demo_blogpage" does not exist
```

**Comment by gasman** [comment]
I haven't tried bringing up the feature/streamfield-frontend branches from scratch, so I can't guarantee that will work... Try running the migrations with the master branch of wagtaildemo checked out, and then switch to the feature/streamfield-frontend branch.

**Comment by edrex** [comment]
Realized i needed to run `\c wagtaildemo` before running the SQL above, no error then.

Still same error running migrate on `feature/streamfield-frontend` branch after running `migrate` and `load_initial_data` from an empty DB with master checked out (both repos).

**Comment by gasman** [comment]
OK, it turns out it is a legitimate error - I haven't set up the new `StreamField` model field (introduced in the feature/streamfield-frontend branch) to work with migrations yet. Here are the steps that work for me, based on running the migrations on feature/streamfield (actually, I think going directly from master to feature/streamfield-frontend should work, as long as you don't try running migrate on feature/streamfield-frontend):

    git clone https://github.com/torchbox/wagtaildemo.git
    cd wagtaildemo
    vagrant up
    git checkout feature/streamfield
    vagrant ssh
    pip install -e git+https://github.com/torchbox/wagtail.git@feature/streamfield#egg=wagtail
    ./manage.py createsuperuser
    ./manage.py migrate
    psql -Upostgres wagtaildemo
	(run the SQL above)
    ./manage.py runserver 0.0.0.0:8000
	(check that the site is running OK)
    git checkout feature/streamfield-frontend
    pip install -e git+https://github.com/torchbox/wagtail.git@feature/streamfield-frontend#egg=wagtail
    ./manage.py runserver 0.0.0.0:8000
	(check that the site is running OK)

**Comment by edrex** [comment]
Ok, home page renders under feature/streamfield. Under feature/streamfield-frontend getting a ValueError rendering the home page:

http://dpaste.com/24RCE55

**Comment by gasman** [comment]
Yes, that's expected - the homepage is trying to output a listing of blog posts, but we've only converted the first blog post page (ID 16, "Blog Post") to StreamField data. The other one still has HTML content from the RichTextField, and it's failing because that isn't valid JSON. To see StreamField in action, you should edit the page "Blog Post".

Bear in mind that this is still a work in progress, and there will be errors :-)

**Comment by davecranwell** [comment]
Interface/Interaction issues outstanding:

- [ ] label on block buttons should allow multiline before ellipsis (or instead of) (css)
- [ ] highlighting of hovered block doesn't stick when block is left e.g when buttons in RTE are hovered. It may be hard to procedurally test whether a structblock, for example, is active. (js/css)
- [ ] + button for item above becomes hovered sometimes when hovering over content of item below. (css)
- [ ] tabbing between field doesn't seem to work at all (js)
- [ ] keyboard shortcuts would be nice, particularly for opening the block menu to create a new block (js)
- [ ] Page/Image/Doc/Snippet choosers need integrating
- [ ] Would be nicer if hallojs toolbar appeared higher than it does, and without the input to which it relates getting top padding. This was added deliberately to make it appear like the toolbar was "in" the textarea but in streamfield the result is an ugly jumping of the text you're looking at. In streamfield it would probably be ok for the toolbar to overlap the content above. (js/css)
- [ ] Adding new ListBlocks should ideally show a blank form or the first item, rather than being required to click "Add another" to get started, which is awkward and mis-worded. (js/python)
- [ ] ListBlocks (nb: not simply blocks that simulate lists) need styling (css)
- [ ] @ollywillans has suggested wagtail ship with various blocks by default. Given that the only place blocks are defined are within the client-specific model, how would we do this? (python)
- [ ] all classes added to streamfield blocks by the implementer (via the `classname` attr) are now prefixed with the namespace ".blockclass-". Wagtaildemo needs updating to include a sample hooks file to demonstrate this via `insert_editor_css`. See also above item: How would we ship with blocks that have css targeting them to make titles larger for example? (css/python)

**Comment by gasman** [comment]
> all classes added to streamfield blocks by the implementer (via the classname attr) are now prefixed with the namespace ".blockclass-". (Might ".userclass-" be better?)

I don't think there should be a prefix on here at all. The `blockname-` prefix makes sense because the user didn't explicitly ask for that classname to be put there, and they should be able to name their blocks whatever they want without having to consider the risk of classname collisions. (Aside: surely it should be `blockname-{{ name }}`, not `blockname-{{ label }}`, because the label can contain spaces, punctuation and capitalisation.) However, when they set a `classname` attribute, they specifically want those classnames to appear on the element, whether that's a Wagtail-provided style like `full title` or one of their own - and that's exactly what we should give them. Adding a prefix is basically just adding an extra layer of obtuseness to say "we don't trust you to choose your classnames responsibly" :-)

**Comment by davecranwell** [comment]
Regarding `blockname` yes that should be `{{ name }}` not `{{ label }}`. 

Regarding prefixing classes supplied via `classname`, we discussed this last week and seemed to be on the same page, but perhaps I explained poorly. I suggested it to avoid clashes but perhaps it is a little too untrusting and clumsy.

**Comment by gasman** [comment]
Ah, I was under the impression that the conversation last week was about the `blockname-{foo}` class in particular, not the others that are provided through `classname`.

**Comment by davecranwell** [comment]
- [ ] Order of content entered doesn't seem to be retained on save. All blocks were remembered, but in a seemingly arbitrary order

**Comment by davecranwell** [comment]
- [ ] Require a means of supporting fields created using blocks that have been subsequently deleted/renamed. At present you get something similar to:

```KeyError at /admin/pages/5/edit/
u'image'
Exception Location:	/home/vagrant/.virtualenvs/globalwitnesswagtail/src/wagtail-master/wagtail/wagtailadmin/blocks.py in to_python, line 805```

**Comment by davecranwell** [comment]
- [ ] Running `makemigrations` on a content type using streamfield will generated new migration files every run, even if nothing has changed.

**Comment by gasman** [comment]
> Order of content entered doesn't seem to be retained on save. All blocks were remembered, but in a seemingly arbitrary order

Can't replicate this. Have tried inserting blocks out-of-order, deleting and then re-inserting, different block types, and various combinations of saving / previewing / publishing - the correct ordering is always preserved for me.

**Comment by gasman** [comment]
> Order of content entered doesn't seem to be retained on save. All blocks were remembered, but in a seemingly arbitrary order

Now identified, and fixed (with test case) in 38b2977.

**Comment by gasman** [comment]
> Running makemigrations on a content type using streamfield will generated new migration files every run, even if nothing has changed.

Fixed in 047ff73. (39 lines of comments for 3 lines of code... that has to be a personal record.)

**Comment by davecranwell** [comment]
Having created several migrations with my blocks defined in a testing content type I now want to abstract my blocks into another python file to be inherited by more content types, but all my migrations refer to their previous location in my  "reports" model e.g

```
migrations.AlterField(
            model_name='reportpage',
            name='streamfield',
            field=wagtail.wagtailcore.fields.StreamField([(b'h2', 
wagtail.wagtailcore.blocks.CharBlock(classname=b'title', template=b'streamfield_blocks/h2.html', icon=b'title')), 
            ... [etc] ...
(b'alignment', reports.models.ImageFormatChoiceBlock())], icon=b'image', label=b'Aligned image')), 
            ... [etc] ...
        ),
```
I've now moved this block to `utils.blocks` so have to rewrite all my migrations. Indeed I had to rewrite all my migrations which overlapped with the move of blocks from wagtailadmin to wagtailcore, but I appreciate that was a one-time change, however moving around blocks in a codebase seems a perfectly likely/common scenario.

This seems likely to be a pain in the neck. It's relatively easily solved with a search and replace in the migrations directory, but I wonder if there's any other way of serialising the block to not create these breakable references.

**Comment by gasman** [comment]
@davecranwell Does this still happen on new migrations created yesterday/today? I updated this logic in 64d5ffe2b28803b09b6804cb42e4437a3f469da1, and the new expected behaviour is for `reports.models.ImageFormatChoiceBlock()` to be deconstructed to an instance of `wagtail.wagtailcore.fields.StreamField` with all of its fields declared inline.

(I've already had to work around one Django bug to make this work, so it's possible that it's not entirely reliable!)

**Comment by davecranwell** [comment]
I'm on the latest version of streamfield, with the following block, found in utils.blocks:

```
class ImageFormatChoiceBlock(FieldBlock):
    field = forms.ChoiceField(choices=(('left','Wrap left'),('right','Wrap right'),('mid','Mid width'),('full','Full width'),))

class PullQuoteBlock(StructBlock):
    quote = CharBlock(classname="quote title")
    attribution = CharBlock()

class StoryBlock(StreamBlock):
    h2 = CharBlock(icon="title", classname="title")
    h3 = CharBlock(icon="title", classname="title")
    h4 = CharBlock(icon="title", classname="title")
    intro = RichTextBlock(icon="pilcrow")
    paragraph = RichTextBlock(icon="pilcrow")
    aligned_image = ImageBlock(label="Aligned image", icon="image")
    pullquote = PullQuoteBlock()
    raw_html = RawHTMLBlock(icon="code", label='Raw HTML')
```

If I then add a new child, `h1 = CharBlock(icon="title", classname="title")`

The resulting migration reads:
```
operations = [
        migrations.AlterField(
            model_name='briefingdocumentpage',
            name='body_text',
            field=wagtail.wagtailcore.fields.StreamField([(b'h1', wagtail.wagtailcore.blocks.CharBlock(classname=b'title', icon=b'title')), (b'h2', wagtail.wagtailcore.blocks.CharBlock(classname=b'title', icon=b'title')), (b'h3', wagtail.wagtailcore.blocks.CharBlock(classname=b'title', icon=b'title')), (b'h4', wagtail.wagtailcore.blocks.CharBlock(classname=b'title', icon=b'title')), (b'intro', wagtail.wagtailcore.blocks.RichTextBlock(icon=b'pilcrow')), (b'paragraph', wagtail.wagtailcore.blocks.RichTextBlock(icon=b'pilcrow')), (b'aligned_image', wagtail.wagtailcore.blocks.StructBlock([(b'image', wagtail.wagtailimages.blocks.ImageChooserBlock()), (b'alignment', utils.blocks.ImageFormatChoiceBlock())], icon=b'image', label=b'Aligned image')), (b'pullquote', wagtail.wagtailcore.blocks.StructBlock([(b'quote', wagtail.wagtailcore.blocks.CharBlock(classname=b'quote title')), (b'attribution', wagtail.wagtailcore.blocks.CharBlock())])), (b'raw_html', wagtail.wagtailcore.blocks.RawHTMLBlock(label=b'Raw HTML', icon=b'code'))]),
            preserve_default=True,
        ),
    ]
```

which looks like the old format to me

**Comment by gasman** [comment]
Ah yes - the decomposition only happens for subclasses of `StructBlock` and `StreamBlock`, so it won't happen on `ImageFormatChoiceBlock`, which is a subclass of `FieldBlock`.

This is impossible to solve in the general case.

(rambly explanation follows, because it's worth getting this documented properly)

For migrations to work properly and not need rewriting retrospectively, there needs to be a clear dividing line between:
1) things that are guaranteed never to change their signature / location throughout the lifetime of the project - which are thus safe to reference within migrations - and
2) things that are allowed to change over time, meaning that migrations have to contain a 'frozen' description of them defined entirely in terms of 1).

Usually, this dividing line will exist at the boundary between `wagtail.wagtailcore.blocks` and your project - i.e. migrations should only contain references to things in `wagtail.wagtailcore.blocks`. We can make this work for subclasses of `StructBlock` and `StreamBlock`, because there's a standard formalised way of subclassing those without writing any custom code. The base classes know about this system, and contain the logic to construct a frozen definition for any subclass you create. 

With `FieldBlock`, there's no such standardisation - subclasses of `FieldBlock` can override any behaviour they like. This means that as soon as you create a subclass of FieldBlock, you're writing custom code that would traditionally belong on the `wagtail.wagtailcore.blocks` side of the fence, and at that point you're committed to keeping that code in place forever.

(end rambly explanation)

The best we can do is define a new `ChoiceBlock` block type within `wagtail.wagtailcore.blocks`, so that `ImageFormatChoiceBlock` can be defined in terms of that - we can then set this up so that it gets decomposed to a reference to `wagtail.wagtailcore.blocks.ChoiceBlock` within migrations.

This would be a worthwhile thing for us to do *if* we think that people will be routinely creating their own multiple-choice block types within  and casually moving them around / renaming them on a whim; that's probably *not* the case for `ImageFormatChoiceBlock` (which will presumably become a stock Wagtail block type once things have settled down), but I can see that situation coming up elsewhere, so I reckon it's worth doing.

**Comment by davecranwell** [comment]
So if I understand rightly this will be encountered any time an implementer wants to use a regular django field which we've not explicitly recreated as a "block" field? I can see uses for Booleanfield and Datefield certainly. Would it be worth adding these while we're at it?

**Comment by gasman** [comment]
@davecranwell Yep - creating your own FieldBlock subclasses is very much in "category 1" above, and we should aim to have a comprehensive set of these in wagtailcore.blocks so that users don't have to do this in their own apps. BooleanField and DateField are simpler cases, as there's no need to customise those for application-specific purposes (they'll be used as-is in definitions) - I'll get these implemented along with ChoiceBlock.

**Comment by davecranwell** [comment]
Is it possible to make the process of transitioning to a streamfield from a non-streamfield a bit easier?

With a current client we've stubbed out models using RichTextField, but now need that field to use Streamfield instead, but our databases contain incompatible data for the field (non-json) and the following line throws errors:
https://github.com/torchbox/wagtail/blob/feature/streamfield/wagtail/wagtailcore/fields.py#L68

Is it terribly yucky to catch this error and ignore it, giving the user a blank field instead?

**Comment by gasman** [comment]
@davecranwell The proper place to deal with that would be in migrations, rather than trying to get StreamField to handle data that wasn't intended for it. However, I'm not sure how smart we can make migrations - since StreamField and RichTextField are using the same data type at the database level (i.e. text), the migration framework will tend to decide that it has no changes to make.

The simple-but-effective way around this would be to delete and recreate the field, in two separate migrations.

Ideally we'd be able to migrate the content, by converting rich text content to a StreamField with a single rich text block (not quite as good as splitting out images etc, but it'll do as a first pass) - but 1) that would need some manual intervention on a per-migration basis, because StreamField itself doesn't know the correct block name to use, and 2) I'm not sure how to incorporate this into the migration process - evidently it would have to happen before the migrations 'see' the change to StreamField, or else we're back to the problem of retrieving data from a StreamField that the StreamField considers to be invalid. Will ponder this some more...

**Comment by gasman** [comment]
Now merged into master.

**Comment by davecranwell** [comment]
Would it make sense to merge wagtaildemo's streamfield branch into its own master too?

**Comment by gasman** [comment]
No, because that'll break it for everyone currently running wagtaildemo master against 0.8.x.

**Comment by gasman** [comment]
...also, a lot of stuff in that branch, like ExpertSpeakerBlock and ShoppingListBlock, are really just experiments to try out concepts of the blocks API - we'll probably want to chuck a lot of that out and make it more coherent before setting end users loose on it.

**Comment by davecranwell** [comment]
Is it worth us creating a "stable/0.8.x" branch on Wagtaildemo too?

**Comment by gasman** [comment]
If anything, we should keep the 'master' branch as stable and add a new 'development' branch for the 0.9 stuff - that way, people using wagtaildemo for demoing don't have to switch branches to get a non-experimental version.

I'm sceptical about how much use a development branch of wagtaildemo will get - it's not as if we've been rushing to update wagtaildemo with pre-release features before now :-) I don't see the problem with letting wagtaildemo lag behind Wagtail master and just updating it once the new Wagtail release is out - but if you think a new branch is worthwhile, go for it.

**Comment by DheerG** [comment]
@gasman Figured I would come out of the shadows for this one. Please don't let the wagtaildemo lag behind Wagtail master especially in the case of important core features being added. I for one would have loved to see a working example of StreamField.

I've been testing wagtail for the fast week for a large project I am working on, the only thing holding us back is getting StreamField working. I currently keep running into:
https://github.com/torchbox/wagtail/blob/feature/streamfield/wagtail/wagtailcore/fields.py#L68

Wagtaildemo on a development branch alone would have most likely greatly helped troubleshoot it.

**Comment by gasman** [comment]
@DheerG Fair enough! I should point out that a 'development' branch wouldn't have helped much in this situation: StreamField was only merged into master two days ago, so it would have lived on a feature branch for all of this time either way - and even now, it's not in a very end-user-friendly format. Currently you still need to apply the database changes to wagtaildemo manually, and you'll still get a JSON decode error if you visit any pages that reference un-migrated BlogPost pages - it wouldn't make sense to create a 'polished' example of StreamField in wagtaildemo while the feature is still in development.

I'll certainly keep it in mind for any completed-but-not-yet-released features in future, though!

**Comment by gasman** [comment]
A bunch of new FieldBlock-based block types now added to `wagtail.wagtailcore.blocks`. All of these support `required` (defaulting to True) and `help_text` as keyword arguments.

* `DateBlock()`, `TimeBlock()`, `DateTimeBlock()` - date/time pickers using our standard JS widgets
* `BooleanBlock()` - a checkbox. Note that this follows Django's standard, somewhat counter-intuitive behaviour where you have to explicitly pass `required=False` to get a checkbox that you can actually turn on and off - without it, it will be an "I agree to the terms and conditions"-type thing that you have to tick to pass validation.
* `ChoiceBlock()` - a select box. Choices can be passed to the constructor:

    `blocks.ChoiceBlock(choices=[('tea', 'Tea'), ('coffee', 'Coffee')], required=False)`

or by subclassing ChoiceBlock:

    class BeverageChoiceBlock(blocks.ChoiceBlock):
        choices = [
            ('tea', 'Tea'),
            ('coffee', 'Coffee'),
        ]

        class Meta:
            icon = 'cup'
            template = 'demo/blocks/beverage.html'

    ....

    BeverageChoiceBlock(required=False)

in which case it will use the same trick as StructBlock / StreamBlock and deconstruct to a plain ChoiceBlock within migrations, so that you don't have a frozen reference to BeverageChoiceBlock.

**Comment by davecranwell** [comment]
Nice one @gasman. Multiline text inputs (textareas) would be a good one next: they're particularly useful for the streamfield interface where we can already make a text input appear borderless to give the illusion of typing magically into a google doc, however without multiline support they always cut off at the far edge.

**Comment by gasman** [comment]
`TextBlock` now added. (Height is a bit excessive due to Django specifying `rows="10"` by default, but presumably that's fixable in CSS...)

**Comment by gasman** [comment]
Block reordering now implemented on lists and streams. @davecranwell: this needs a bit of CSS fixing - block type labels sometimes overlap the buttons, and the buttons probably shouldn't collapse entirely when disabled (because we then end up with the bottom-most 'move up' arrow in a list aligned with all of the 'move down' arrows).

**Comment by sixpearls** [comment]
Guys, this is fantastic. I have another couple weeks off and came back to play my favorite web project, Wagtail. I got the demo running and played with the StreamField. It's fantastic! I need to keep digging through the source code, but there's a couple questions/thoughts I have:

- Is it possible to have a block render without a ``<div>`` element around it? Since the structure is semantic, I would like to have the the rendered HTML be semantic as well

- I need to start piecing together the source code i.e., the discussion from #1067, but is there a reason each StreamField definition needs to explicitly define every type of allowed block? The ``ModelAdmin`` or ``Form`` pattern seems like it could be a good choice here. A hook allows each app to register their own blocks. Then the creation of the StreamField with no arguments allows all registered blocks to be used. An ``include`` argument limits the blocks to those explicitly defined, and an ``exclude`` argument can be used to explicitly prevent blocks from being used.

Is #1033 still "available"? I'd love to try to get in a commit or two before classes start back up again.

**Comment by gasman** [comment]
Hi @sixpearls - thanks for the kind words!

* The `<div>` markup is just the default rendering you get when outputting the field as a single value (as in `{{ self.body }}`), in much the same way that Django forms have a not-very-pretty native rendering. For more control over the output, you can iterate over it to get a sequence of `StreamChild` objects that can be rendered individually, and inspected through the `block_type` and `value` properties:

    {% for block in self.body %}
        {% if block.block_type == 'heading' %}
            <h1>{{ block.value }}</h1>
        {% else %}
            {{ block }}
        {% endif %}
    {% endfor %}

* If you've got a standard set of blocks that's used across StreamFields on various page models, your best option is to subclass StreamBlock:

    class StoryBlock(blocks.StreamBlock):
        heading = blocks.CharBlock()
        paragraph = blocks.RichTextBlock()
        photo = ImageChooserBlock()

You can then pass `StoryBlock()` as the parameter to StreamField, rather than the list of blocks. Early on in development I considered the idea of having block types registered globally, but I felt that there was potential for the 'stream of different object types' mechanism to be used in several separate contexts - such as carousels, or the form builder - and having a single 'master' set of blocks would have worked against this.

Putting both of these answers together, you can set up a subclass of StreamBlock to implement its own rendering to be used by `{{ self.body }}` in place of the default `<div>`-based output - either by defining a `render` method, or (more straightforwardly) setting a `template` property in Meta:

    class StoryBlock(blocks.StreamBlock):
        heading = blocks.CharBlock()
        paragraph = blocks.RichTextBlock()
        photo = ImageChooserBlock()

        class Meta:
            template = "myapp/blocks/story_block.html"

(Within the template, the stream object as a whole will be referred to as `self`.)

#1033 hasn't been claimed, so you're very welcome to have a go at that! I reckon the starting point for that would be wagtailadmin/static/wagtailadmin/js/blocks/sequence.js - you'd need to supplement the moveUp / moveDown logic with a new method to allow moving items to an arbitrary position, rather than just up or down one step.

**Comment by sixpearls** [comment]
@gasman Ah, I wasn't really considering different _types_ of StreamFields. I'll need to give it a think on how I would implement this into a project, but I am beginning to see what you're trading off. I still have those small blogs to churn out; I'll reserve anymore critiques until I've implemented a site using the new field. Thanks for the tips on how to get started!

I'll try to get a prototype out for #1033. I also wanted to play with the richtext block and hallo some more. I'll PR anything that seems worth while.

**Comment by zerolab** [comment]
Should this be closed now? StreamField is in master, docs have been merged, there is a release note..

**Comment by gasman** [comment]
I'm not counting it as complete until the rendering bugs (e.g. overlapping labels and up/down arrows) are solved. But I guess that can be created as a separate ticket...

**Comment by gasman** [comment]
As per @zerolab's suggestion, will close this now and open new tickets for the remaining work.
